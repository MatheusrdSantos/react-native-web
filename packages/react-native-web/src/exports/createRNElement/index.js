/**
 * Copyright (c) Nicolas Gallagher.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import React from 'react';
import { unstable_createElement as createElement } from '../createElement';
import useElementLayout from '../../hooks/useElementLayout';
import usePlatformMethods from '../../hooks/usePlatformMethods';
import useResponderEvents from '../../hooks/useResponderEvents';
import setAndForwardRef from '../..//modules/setAndForwardRef';
import pick from '../../modules/pick';

const createRNElement = (tagname, forwardPropsList, defaultStyle) => {
  return React.forwardRef((props, ref) => {
    const {
      onLayout,
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture,
      style
    } = props;
    const hostRef = React.useRef(null);
    const setRef = setAndForwardRef({
      getForwardedRef: () => ref,
      setLocalRef: hostNode => {
        hostRef.current = hostNode;
      }
    });

    const classList = [defaultStyle];
    useElementLayout(hostRef, onLayout);
    usePlatformMethods(hostRef, { classList, style });
    useResponderEvents(hostRef, {
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture
    });
    const supportedProps = React.useMemo(() => {
      const finalProps = pick(props, forwardPropsList);
      finalProps.classList = classList;
      finalProps.ref = setRef;
      finalProps.style = style;
      return finalProps;
    }, [props, classList, setRef, style]);
    return createElement(tagname, supportedProps);
  });
};

export default createRNElement;
