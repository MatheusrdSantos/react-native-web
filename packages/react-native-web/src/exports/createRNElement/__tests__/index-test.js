/* eslint-env jasmine, jest */

import createRNElement from '../index';
import React from 'react';
import { render, fireEvent } from '@testing-library/react';

describe('exports/createRNElement', () => {
  test('renders different DOM elements', () => {
    const Text = createRNElement('h3', { ref: true });
    const { container: h3Container } = render(<Text>test</Text>);
    expect(h3Container.firstChild).toMatchSnapshot();
    const Button = createRNElement('button', { ref: true });
    const { container: btnContainer } = render(<Button />);
    expect(btnContainer.firstChild).toMatchSnapshot();
  });

  test('renders default style', () => {
    const View = createRNElement('div', {}, { backgoundColor: 'red' });
    const { container } = render(<View style={{ height: 10, width: 10 }} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('prop "href"', () => {
    const propList = {
      ref: true,
      href: true
    };
    const Text = createRNElement('a', propList);
    const { container } = render(<Text href={'http://mylink.com'}>test</Text>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('onPress callback', () => {
    const Button = ({ onPress }) => {
      const Custom = createRNElement('button', { onClick: true });
      return <Custom onClick={onPress} />;
    };
    const myFn = jest.fn();
    const { container } = render(<Button onPress={myFn} />);
    fireEvent.click(container.firstChild);
    expect(myFn).toHaveBeenCalledTimes(1);
  });

  test('onLayout callback', () => {
    const myFn = jest.fn();
    const View = createRNElement('div');
    const { rerender } = render(<View onLayout={myFn} style={{ height: 10, width: 10 }} />);
    rerender(<View onLayout={myFn} style={{ height: 15, width: 15 }} />);
    expect(myFn).toHaveBeenCalledTimes(2);
  });
});
