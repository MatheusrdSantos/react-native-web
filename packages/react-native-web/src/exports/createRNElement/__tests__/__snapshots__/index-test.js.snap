// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports/createRNElement prop "href" 1`] = `
<a
  class="css-reset-4rbku5 9"
  data-focusable="true"
  href="http://mylink.com"
/>
`;

exports[`exports/createRNElement renders default style 1`] = `
<div
  class="8"
  style="height: 10px; width: 10px;"
/>
`;

exports[`exports/createRNElement renders different DOM elements 1`] = `
<h3
  class="6"
/>
`;

exports[`exports/createRNElement renders different DOM elements 2`] = `
<button
  class="css-reset-4rbku5 7"
  data-focusable="true"
/>
`;
